Fix last message preview for Tasks in LNH when completing and reopening a task

PauloGasparSv opened the pull request:
### Details
When you complete or re-open a task the message preview for that Task's report in the LNH was displaying the wrong message sometimes (showing the previous one or empty).

This P.R. **adds optimistic data and failure data** to some Tasks functions and change the logic to use `ReportUtils.getReportName` to get the Task's title. 

### Fixed Issues
https://github.com/Expensify/App/issues/19098
PROPOSAL: GH_LINK_ISSUE(COMMENT)


### Tests

1. Sign into newDot and start a chat with someone.
2. Create a new Task and be sure to share it in the new chat you created.
3. Open that new Task's Report's Chat and click on the "Mark as Done" green button on the top of the page
4. In the LNH, verify that the message preview for that Task's chat shows `completed task: <Task name>`
    <img width="150" alt="image" src="https://github.com/Expensify/App/assets/6564265/074f2eee-5017-4557-a6cf-65c1c56198f2">
5. Inside the Task's Chat, re-open the Task by clicking on the three dots icon and then on "Mark as incomplete"
6. In the LNH, verify that the message for that Task's chat preview shows `reopened task: <Task name>`
    <img width="150" alt="image" src="https://github.com/Expensify/App/assets/6564265/84779c11-126f-40a1-9028-fb4636196e01">
6. Refresh the page and make sure the LNH preview is still correct
7. Inside the Task, let's mark it as completed again but this time using the "Mark as done" option found by clicking on the Three dots menu.
8.  In the LNH, verify that the message preview for that Task's chat shows `completed task: <Task name>`
9. Open the chat you shared that Task. There should be a message with checklist with the Task you created. Re-open the Task by clicking on the Task's checkbox to "unmark it as done":
    <img width="200" alt="image" src="https://github.com/Expensify/App/assets/6564265/a80ea779-0122-46e9-9845-7b19bfa4ba31">
10. In the LNH, verify that the message for that Task's chat preview shows `reopened task: <Task name>`
11. Still on the chat you shared the Task, click on the Task's checkbox again to mark it as done
12. In the LNH, verify that the message preview for that Task's chat shows `completed task: <Task name>`


- [ ] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

Do the following 3 steps and then repeat the same Test steps for online

1. Sign into NewDot
2. Start a new chat with someone
3. Toggle connectivity off

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
4. Verify a modal appears displaying a preview of that image
--->

Same as tests

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->


- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I ran the tests on **all platforms** & verified they passed on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.


### Screenshots/Videos
<details>
<summary>Web</summary>


</details>

<details>
<summary>Mobile Web - Chrome</summary>


</details>

<details>
<summary>Mobile Web - Safari</summary>


</details>

<details>
<summary>Desktop</summary>


</details>

<details>
<summary>iOS</summary>


</details>

<details>
<summary>Android</summary>


</details>

Commit:
Add optimistic action text and data to completeTask

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-complete-task-lnh-last-message-fix

Commit:
Add last message text failure data to completeTask

PauloGasparSv commented:
I've added `lastMessageText` failure data to the complete task call and now I'm trying to fix the following bug:

#### Marking task as Done or Re-opening shows  `reopened task: <email>` instead of task name
1. Assign someone a Task
2. Mark the Task as Done then Re-open
3. The message preview in the LNH should always mention the correct Task title but in some instances it shows an e-mail

I think that happens after I refresh the page and the report is fetched (but it doesn't happen if you create the Task and don't refresh).
This is happening when `getReportName` returns one of the participants name/email as the report's title [here](https://github.com/Expensify/App/blob/d5374be5435eea4b94978f93588ebad8ffdc0867/src/pages/home/HeaderView.js#L80).

![image](https://github.com/Expensify/App/assets/6564265/62548c9d-f441-4e9b-b796-1be924edf86a)

Investigating how to fix this!

PauloGasparSv commented:
Interesting, I think that only happens when I re-open a Task but never when I mark one as done.

Commit:
Use report name as title for completing and reopening Tasks

PauloGasparSv commented:
Ok, fixed the last bug mentioned above and the feature seems to be working!

I'll write test steps and fill the checklist then make this ready for reviews.

PauloGasparSv commented:
I don't see a reason to test this against a high traffic staging/production account so checking that off without doing that.

If anyone sees this and disagrees please LMK!

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-complete-task-lnh-last-message-fix

PauloGasparSv commented:
Hey @danieldoglas added you here by mistake so un-assigning you, I was trying to add @s77rt  here

s77rt commented on src/libs/actions/Task.js:
```diff
@@ -200,6 +207,7 @@ function completeTask(taskReportID, taskTitle) {
             value: {
                 stateNum: CONST.REPORT.STATE_NUM.OPEN,
                 statusNum: CONST.REPORT.STATUS.OPEN,
+                lastMessageText: taskLastMessageText,
```
Don't we need the same logic for CreateTask, ReopenTask, EditTask and CancelTask. All seems to set the `lastMessageText` but never reset it in case of failure. Or would you rather handle this separately?

PauloGasparSv replied:
> Or would you rather handle this separately?

That was the case! But I didn't stop to think about this again after I came up with this solution, since it's pretty simple I'll see if we can do the same for the other calls : )

PauloGasparSv replied:
Added that to ReopenTask, EditTask and CancelTask (but not to CreateTask)

PauloGasparSv replied:
Oh, I think I added a bug in this last commit!

PauloGasparSv replied:
All right, I fixed it! Re-tested and things seem ok

s77rt commented on src/pages/home/HeaderView.js:
```diff
@@ -92,7 +92,7 @@ const HeaderView = (props) => {
                 icon: Expensicons.Checkmark,
                 text: props.translate('newTaskPage.markAsDone'),
-                onSelected: () => Task.completeTask(props.report.reportID, title),
+                onSelected: () => Task.completeTask(props.report.reportID, props.report.reportName, props.report.lastMessageText),
```
```suggestion
                onSelected: () => Task.completeTask(props.report.reportID, title, props.report.lastMessageText),
```
I don't see why this change is required, `title` is not necessary `reportName`.

PauloGasparSv replied:
More info on why I changed that [here](https://github.com/Expensify/App/pull/20373#issuecomment-1585071638)
 
> title is not necessary reportName

I think we can assume the `title` is always the report name for Tasks specifically [because of this logic](https://github.com/Expensify/App/blob/main/src/libs/ReportUtils.js#L995-L997), no?

s77rt replied:
> I think we can assume the title is always the report name

Yes, I think so, but this does not explain why you we are changing the title if we expect them to be the same.

I think the bug you linked is due the passing the wrong report to `getReportName` we are passing the parent report which is a chat report thus the name is the participant's email/display name. I think we may need to pass the report itself (task report).

PauloGasparSv replied:
> I think the bug you linked is due the passing the wrong report to getReportName we are passing the parent report which is a chat report thus the name is the participant's email/display name. I think we may need to pass the report itself (task report).

I agree! Pretty sure I tried to do that and failed but **I'll give another look and try again**!

PauloGasparSv replied:
I'll take some more time to follow up to this, after I merged `main` a bunch of errors started showing for me (when opening the app, reports and creating tasks which says `There was an error assigning this task, please try another assignee.Confirm task` forever)

I'll shift focus to another P.R. then get back to this tomorrow.

bondydaa replied:
`title` can be different based on this logic https://github.com/Expensify/App/blob/926277a188c9952ec567ab47254b1c6e45a2de38/src/pages/home/HeaderView.js#L79-L80

It looks like maybe we have some dup logic when setting `reportHeaderData` since `ReportUtils.getReportName` also does checks for task reports but i'm not 100% sure, just what i'm seeing from a quick look.

s77rt commented on src/pages/home/HeaderView.js:
```diff
@@ -101,7 +101,7 @@ const HeaderView = (props) => {
                 icon: Expensicons.Checkmark,
                 text: props.translate('newTaskPage.markAsIncomplete'),
-                onSelected: () => Task.reopenTask(props.report.reportID, title),
+                onSelected: () => Task.reopenTask(props.report.reportID, props.report.reportName),
```
```suggestion
                onSelected: () => Task.reopenTask(props.report.reportID, title),
```

PauloGasparSv replied:
Same as [here](https://github.com/Expensify/App/pull/20373/commits/21547806c73db98288b7a40ca8808f8b325657ea#r1230112181)

Commit:
Add last message text failure data to cancel and reopen task

Commit:
Fix typo in HeaderView.js

PauloGasparSv commented:
Bump @bondydaa if you could give this a review asap too : )
We were hoping [this could be deployed on Thursday](https://expensify.slack.com/archives/C04QEB4MJEQ/p1686594346644159?thread_ts=1686538027.572239&cid=C04QEB4MJEQ)

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-complete-task-lnh-last-message-fix

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-complete-task-lnh-last-message-fix

Commit:
Pass report object to reopenTask and generate title using getReportName

Commit:
Pass report object to cancelTask too and add remaining failure data

Commit:
Remove duplicate lastMessageText from cancelTask

Commit:
Pass report object to completeTask and add more optimistic and failure data

PauloGasparSv commented:
I changed the logic of `completeTask`,`reopenTask` and `cancelTask` and added failure data to all properties modified by the optimistic. Also added some more Optimistic data to `completeTask` so it matches the other methods.

I also changed these methods to receive the entire `report` object instead of many fields separately. Because of that I added a call to `ReportUtils.getReportName` on each method to get the correct Task's report title. I think that fixes pretty much every point raised in the last review so I'll re-test everything and ask for a re-review!

Commit:
Remove unused variable

PauloGasparSv commented on src/libs/actions/Task.js:
I added the missing failure data of some of the methods! More info [here](https://github.com/Expensify/App/pull/20373#issuecomment-1601571284).

Commit:
Lint TaskPreview and Task

PauloGasparSv commented:
I think there is 1 problem left that I'm trying to replicate where sometimes if you Mark a Task as closed then both the `CompleteTask` and `OpenReport` requests are made pretty much one after the other and the `OpenReport` one returns an empty `lastMessageHtml` and `lastMessageTask`. Because of that and the order requests are made sometimes the message preview becomes empty.

Trying to figure out what causes that before asking for a re-review : )

PauloGasparSv commented:
Since I believe [this bug](https://github.com/Expensify/App/pull/20373#issuecomment-1601658147) is in the API side this is ready for re-reviews @bondydaa @s77rt! 

I'll add evidences tomorrow morning but do you mind giving this a review+test already?

Commit:
Add lastMessageHtml to optimistic and failure data in Task actions

s77rt commented on src/libs/actions/Task.js:
```diff
@@ -186,28 +186,31 @@ function createTaskAndNavigate(currentUserEmail, currentUserAccountID, parentRep
                 stateNum: CONST.REPORT.STATE_NUM.SUBMITTED,
                 statusNum: CONST.REPORT.STATUS.APPROVED,
                 lastMessageText: message,
+                lastMessageHtml: message,
```
```suggestion
                lastMessageHtml: _.escape(message),
```
html and text are never the same. Here we should escape the text in case it contains html entities like `<` or `&`

PauloGasparSv replied:
Not in this specific case right?  Since the `message` here is created on line `192` (similar to the 2 other methods) I think it doesn't contain any html so both 

I think that adding that `_.escape` has no cons so I'll test and add it : )

(same for the other 2 functions)

s77rt replied:
The message contains a dynamic report name. If the report name is e.g. `Day & Night` then this would cause a problem because the html version of this text should be `Day &amp; Night` and not `Day & Night`.

PauloGasparSv replied:
Yeah makes total sense, added that to all the `lastMessageHtml` I've added!

s77rt commented on src/libs/actions/Task.js:
```diff
@@ -242,31 +250,30 @@ function completeTask(taskReportID, taskTitle, originalLastMessageText) {
                 statusNum: CONST.REPORT.STATUS.OPEN,
-                lastVisibleActionCreated: reopenedTaskReportAction.created,
                 lastMessageText: message,
+                lastMessageHtml: message,
```
```suggestion
                lastMessageHtml: _.escape(message),
```

PauloGasparSv replied:
Done!

s77rt commented on src/libs/actions/Task.js:
```diff
@@ -570,39 +582,36 @@ function getShareDestination(reportID, reports, personalDetails) {
             value: {
                 lastVisibleActionCreated: optimisticCancelReportAction.created,
                 lastMessageText: message,
+                lastMessageHtml: message,
```
```suggestion
                lastMessageHtml: _.escape(message),
```

PauloGasparSv replied:
Done!

s77rt commented on src/libs/actions/Task.js:
```diff
@@ -216,24 +219,29 @@ function completeTask(taskReportID, taskTitle, originalLastMessageText) {
-            key: `${ONYXKEYS.COLLECTION.REPORT}${taskReportID}`,
+            key: `${ONYXKEYS.COLLECTION.REPORT}${report.reportID}`,
             value: {
                 stateNum: CONST.REPORT.STATE_NUM.OPEN,
                 statusNum: CONST.REPORT.STATUS.OPEN,
```
```suggestion
                stateNum: report.stateNum,
                statusNum: report.statusNum,
```
I guess we can do that, right?

PauloGasparSv replied:
I think it makes sense (specially if we add other states eventually)! 

Changed that!

s77rt commented on src/libs/actions/Task.js:
```diff
@@ -275,32 +282,37 @@ function reopenTask(taskReportID, taskTitle, originalLastMessageText) {
-            key: `${ONYXKEYS.COLLECTION.REPORT}${taskReportID}`,
+            key: `${ONYXKEYS.COLLECTION.REPORT}${report.reportID}`,
             value: {
                 stateNum: CONST.REPORT.STATE_NUM.SUBMITTED,
                 statusNum: CONST.REPORT.STATUS.APPROVED,
```
```suggestion
                stateNum: report.stateNum,
                statusNum: report.statusNum,
```

PauloGasparSv replied:
Done!

s77rt commented on src/libs/actions/Task.js:
```diff
@@ -216,24 +219,29 @@ function completeTask(taskReportID, taskTitle, originalLastMessageText) {
+            taskReportID: report.reportID,
             completedTaskReportActionID: completedTaskReportAction.reportActionID,
         },
         {optimisticData, successData, failureData},
```
NAB and not a change that was introduced by this PR but I think we can handle it while at it; successData is an empty array, let's remove it and not define it (do same for ReopenTask and EditTask) (you can refer to cancelTask where it's done right)

PauloGasparSv replied:
Removed those empty `successData`

s77rt reviewed (changes requested):
Code looks good just minor requests

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-complete-task-lnh-last-message-fix

Commit:
Escape task last message html in complete, reopen and cancel logic

Commit:
Use original report state data for failureData for task actions

Commit:
Remove empty successData arrays from Task.js

PauloGasparSv commented:
Created https://github.com/Expensify/Web-Expensify/pull/37929 to fix [this](https://github.com/Expensify/App/pull/20373#issuecomment-1601658147) issue in the API by returning the `lastMessageText` in the onyxData for complete, reopen and cancel task.  

Also addressed the comments here and will re-test + add evidences asap

s77rt commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [X] I verified the correct issue is linked in the `### Fixed Issues` section above
- [X] I verified testing steps are clear and they cover the changes made in this PR
    - [X] I verified the steps for local testing are in the `Tests` section
    - [X] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [X] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I verified tests pass on **all platforms** & I tested again on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [X] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [X] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/1528e055-cc23-4ee7-817c-5cf21b9b13e2


</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/7f7fd8c8-21c9-49de-8404-bb3a76efe499


</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/db3bd885-cfab-43b3-90e5-20d8d32bb469


</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/cd4052a1-ccba-4fc3-9ffa-88abb5506757


</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/eff70b4e-c5c3-4a89-be49-857527ad6e66


</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/8bb92c9e-6dea-4c5e-96d5-0711040d4890


</details>

s77rt commented:
Code looks good and tests well.

@PauloGasparSv Can you please complete the checklist and tag me again?

bondydaa commented on src/components/ReportActionItem/TaskPreview.js:
```diff
@@ -74,13 +77,7 @@ function TaskPreview(props) {
-                            TaskUtils.completeTask(props.taskReportID, taskTitle);
-                        }
-                    }}
+                    onPress={() => (isTaskCompleted ? TaskUtils.reopenTask(props.taskReport) : TaskUtils.completeTask(props.taskReport))}
```
this might be a good spot to use the hook `useCallback` https://react.dev/reference/react/useCallback

`isTaskCompleted` and `taskReport` would be the dependencies, i think

PauloGasparSv replied:
Cool, I'll check that in a bit! Thks for the suggestion

s77rt replied:
I think it will be a waste in this case. The Checkbox will always re-render since we don't use `React.memo`.

If we need to optimise this we will have first to optimise Checkbox to use `React.memo` using a custom [`arePropsEqual`](https://react.dev/reference/react/memo) callback, e.g.:
```js
export default React.memo(Checkbox, (prevProps, nextProps) => {
    if (prevProps.isChecked !== nextProps.isChecked) {
        return false;
    }
    if deep compare style not equal return false;
    if deep compare containerStyle not equal return false;
    etc.
    return true;
});
```

bondydaa commented on src/pages/home/HeaderView.js:
same here these methods could suffer from needless re-renders so `useCallback` might be beneficial but I'm not 100% confident here.

PauloGasparSv replied:
Will check this too!

@s77rt please lmk your opinion on this : )

s77rt replied:
Hmm, I think same note above applies here. I didn't check in depth but I doubt this will save us any re-renders.

> these methods could suffer from needless re-renders

We don't use `useCallback` to prevent function re-declaration (assuming that's what you meant). AFAIK it's cheap to re-declare a function over and over. We use `useCallback` to prevent re-renders, as a rule of thumb, use `useCallback` if:
1. The function is being used as dependency (in  `useEffect`, `useMemo`, etc.)
2. The function is being passed to a component that uses it as a dependency
3. The function is being passed to a component that uses memorisation

bondydaa commented on src/libs/actions/Task.js:
```diff
@@ -398,7 +417,7 @@ function editTaskAndNavigate(report, ownerEmail, ownerAccountID, {title, descrip
             assigneeChatReportActionID: optimisticAssigneeAddComment ? optimisticAssigneeAddComment.reportAction.reportActionID : 0,
         },
-        {optimisticData, successData, failureData},
+        {optimisticData, failureData},
```
Are we sure we don't need to pass successData at all here? I thought I remember something like if successData didn't need to be different than optimisticData we should still pass `successData: optimisticData`?

PauloGasparSv replied:
There are other calls to API.write that omit the `successData` and after testing everything worked fine so I don't think we need to pass that.

Since I'm not 100% sure let's ask Jack since he introduced these specific lines.

@thienlnam, do you remember if there is a reason for empty array `successData` and if it's ok to remove it?

thienlnam replied:
Ah yeah the empty successData is okay to remove

> I thought I remember something like if successData didn't need to be different than optimisticData we should still pass successData: optimisticData?

We don't do this because the data is set optimistically on the device, and then on success it sends pusher updates to all clients to update the onyxData. So in this case, if we had successData, it would merge the same data as optimisticData on success, but then we are also getting the pusher data which is the same as the optimisticData so nothing really changes

bondydaa replied:
coolio thanks for the clarification

bondydaa reviewed (approved):
I'm not super confident in my reviews here, we could probably merge as is and add the hooks if we see performance regressions.

PauloGasparSv commented:
I'll address the comments and add evidences tomorrow!

chiragsalian commented on src/libs/actions/Task.js:
```diff
@@ -184,113 +184,133 @@ function createTaskAndNavigate(currentUserEmail, currentUserAccountID, parentRep
+ * @param {Object} report
+ */
+function completeTask(report) {
+    const message = `completed task: ${ReportUtils.getReportName(report)}`;
```
this looks odd to me, i don't think we should save "completed task:" to the message. It's not translated. Instead, i think the component should use another key to know if the task is completed or not and then translate and show the appropriate text. Thoughts?

PauloGasparSv replied:
Makes total sense, I had a [similar discussion](https://expensify.slack.com/archives/C04QEB4MJEQ/p1687385149610769) about this on the API side but didn't think of changing these lines 🤦‍♂️

I think we can use [this](https://github.com/Expensify/App/blob/c122266f96bab047cda55f0cebff8b3a61e97700/src/languages/en.js#L1292-1294) to localize that message so I'll try doing that

bondydaa commented:
what's the status here?

PauloGasparSv commented:
> what's the status here?

Did not have time to focus on this [since this](https://github.com/Expensify/App/pull/20373#issuecomment-1605079467), I'm still focused on a breaking WAQ issues + Manual Requests + https://github.com/Expensify/Expensify/issues/296893 then will have time for this

chiragsalian commented:
Removing myself as reviewer since i'll be on parental leave for a bit.

bondydaa commented:
closing this out

s77rt commented:
@PauloGasparSv @bondydaa What's the next steps here? We had this issue https://github.com/Expensify/App/issues/22059 hold for that PR. Should we just remove the hold on continue working on it there?
